site/public/js/rainbow-customization.js

```diff
@@ -267,78 +267,144 @@ function getGradeableBuckets() {
+/**
+ * Adds input data to plagiarism and manual grade tables
+ * @param {int} table
+ *      0 -> plagiarism table
+ *      1 -> manual grading table
+ */
+function addToTable(table) {
+    if (table !== 0 && table !== 1) {
         return;
     }
 
+    const tableMap = {
+        0: ['plagiarism-table-body', 'plagiarism_user_id', 'g_id', 'marks'],
+        1: ['manual-grading-table-body', 'manual_grading_user_id', 'manual_grading_grade', 'manual_grading_note'],
+    };
+
+    const firstInput = document.getElementById(tableMap[table][1]).value.trim();
+    const secondInput = document.getElementById(tableMap[table][2]).value.trim();
+    const thirdInput = document.getElementById(tableMap[table][3]).value.trim();
+
+    // Check whether input is allowed
     // eslint-disable-next-line no-undef
     const studentFullDataValues = studentFullData.map((item) => item.value);
-    if (!studentFullDataValues.includes(USERID)) {
-        alert('Invalid User ID. Please enter a valid one.');
-        return;
-    }
+    const tableBody = document.getElementById(tableMap[table][0]);
+    const rows = tableBody.getElementsByTagName('tr');
+    switch (table) {
+        case 0: { // plagiarism table
+            if (firstInput === '' || secondInput === '' || thirdInput === '') {
+                alert('Please fill in all the fields.');
+                return;
+            }
+            if (!studentFullDataValues.includes(firstInput)) {
+                alert('Invalid User ID. Please enter a valid one.');
+                return;
+            }
+            if (thirdInput > 1 || thirdInput < 0) {
+                alert('Penalty must be between 0 - 1');
+                return;
+            }
 
-    // Check for penalty
-    if (penalty > 1 || penalty < 0) {
-        alert('Penalty must be between 0 - 1');
-        return;
-    }
+            // Check for duplicate entries
+            for (let i = 0; i < rows.length; i++) {
+                const row = rows[i];
+                const existingFirstInput = row.cells[0].textContent.trim();
+                const existingSecondInput = row.cells[1].textContent.trim();
 
-    const tableBody = document.getElementById('table-body');
+                if (firstInput === existingFirstInput && secondInput === existingSecondInput) {
+                    alert('Entry with the same Student ID and Gradeable already exists.');
+                    return;
+                }
+            }
+            break;
+        }
+        case 1: { // manual grading table
+            if (firstInput === '' || secondInput === '') {
+                alert('Please fill in both user ID and final grade.');
+                return;
+            }
+            if (!studentFullDataValues.includes(firstInput)) {
+                alert('Invalid User ID. Please enter a valid one.');
+                return;
+            }
+            const allowed_grades = ['A', 'A-', 'B+', 'B', 'B-', 'C+', 'C', 'C-', 'D+', 'D', 'F'];
```

Comment:
This set of grades is being reused in multiple places.  Is there an easy way to share code?

Reply:
[a6f1920](https://github.com/Submitty/Submitty/pull/10653/commits/a6f19203f20d9f9cab94c59d71a4efabf705e9dd)